sandbox/cgroup/tracking.go

```diff
@@ -284,6 +296,63 @@ var doCreateTransientScope = func(conn *dbus.Conn, unitName string, pid int) err
 		properties,
 		aux,
 	)
+	var wg sync.WaitGroup
+	closeChan := make(chan struct{})
+	defer func() {
+		close(closeChan)
+		wg.Wait()
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		jobResults := make(map[dbus.ObjectPath]string, 10)
+		expectedJob := dbus.ObjectPath("")
+		for {
+			select {
+			case job, ok := <-jobWaitFor:
+				if !ok {
+					continue
+				}
+				if result, ok := jobResults[job]; ok {
+					// maybe we already have result for this job
+					jobResultChan <- result
+				} else {
+					expectedJob = job
+				}
+			case sig, ok := <-signals:
+				if !ok {
+					continue
+				}
+				// sanity check the signal name, although the
+				// match selectors should ensure we only receive
+				// JobRemoved signals
+				if sig.Name != "org.freedesktop.systemd1.Manager.JobRemoved" {
+					continue
+				}
+				var id uint32
+				var jobFromSignal dbus.ObjectPath
+				var unit string
+				var result string
+				if err := dbus.Store(sig.Body, &id, &jobFromSignal, &unit, &result); err != nil {
+					continue
+				}
+				if jobFromSignal == expectedJob {
+					// we are already expecting results for this job
+					jobResultChan <- result
+				} else {
+					// or not, just keep result for now, as
+					// a request to track a job may come
+					// later
+					jobResults[jobFromSignal] = result
+				}
+			case <-closeChan:
+				conn.RemoveSignal(signals)
+				conn.RemoveMatchSignal(jobRemoveMatch...)
```

Comment:
Are you sure it's fine to manipulate the connection from two different threads? Maybe these two operations should be moved to the parent function, after the signal is received? (maybe even in a `defer` block?)

Reply:
Yes, there's locking in place and IIRC messages are passed around through channels and thus serialized.